docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+
+## Goal
+
+With third party API servers, we want to enable cluster admins or users to
+expose new APIs at runtime without having to modify the core kubernetes API
+server binary.
+Anyone can write their own API server and register it with the core
+kubernetes API server. These new APIs will be seamless extension of core
+kubernetes APIs (ex: they will be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+Users can bring up their own API servers and register them with the core server
+at runtime.
+Core server will then start exposing these third party servers in the discovery
+API. kubectl will use the discovery API to connect to these servers.
+
+These API servers will have full control of their resources.
+Each server will have their own storage mechanism and will implement their own
+authorization and authentication.
+
+
+## Implementation Details
+
+### Thirdpartyapiservers Extensions Resource
+
+API server will support a non-namespaced `thirdpartyapiservers` extension resource.
+Third party servers can POST at
+`http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to register themselves.
+Clients can GET `http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to get a list of registered third party servers.
+The API definition will be as follows:
+
+```
+// ThirdPartyAPIServer represents the configuration of a third party server.
+type ThirdPartyAPIServer struct {
+  unversioned.TypeMeta
+
+  ObjectMeta
+
+  Spec ThirdPartyAPIServerSpec
+
+  Status ThirdPartyAPIServerStatus
+}
+
+// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
+// resource.
+type ThirdPartyAPIServerSpec struct {
+  // Base path of the API server.
+  BasePath string
+
+  // Path to the swagger spec.
+  SwaggerSpecPath string
+}
+
+type ThirdPartyAPIServerStatus struct {
+  Conditions []ThirdPartyAPIServerCondition
+}
+
+type ThirdPartyAPIServerConditionType string
+const (
+        ThirdPartyAPIServerValid ThirdPartyAPIServerConditionType = "Valid"
+)
+
+type ThirdPartyAPIServerCondition struct {
+  Type ThirdPartyAPIServerConditionType
+  // Status of the condition, one of True, False or Unknown.
+  Status ConditionStatus
+  Reason string
+  Message string
+}
+```
+
+There are a few requirements that the third party API servers should meet
+(for ex: name of resources provided by each server should be unique).
+ThirdPartyAPIServerController will look for new servers and will validate them.
+It will set the valid condition in the status accordingly.
+Note that the server is validated only once so it is possible to change it after
+the validation is done. We expect that the server will be deleted and added
+again, whenever its functionality is changed.
+
+### kubectl
+
+Users will be able to use resources provided by third party servers seamlessly
```

Comment:
One problem with describe


On Nov 16, 2015, at 1:46 PM, Daniel Smith <notifications@github.com> wrote:

In docs/proposals/third-party-api-servers.md
<https://github.com/kubernetes/kubernetes/pull/17191#discussion_r44963839>:

> +  Reason string
> +  Message string
> +}
> +```
> +
> +There are a few requirements that the third party API servers should meet
> +(for ex: name of resources provided by each server should be unique).
> +ThirdPartyAPIServerController will look for new servers and will validate them.
> +It will set the valid condition in the status accordingly.
> +Note that the server is validated only once so it is possible to change it after
> +the validation is done. We expect that the server will be deleted and added
> +again, whenever its functionality is changed.
> +
> +### kubectl
> +
> +Users will be able to use resources provided by third party servers seamlessly

I was considering in another context that it might be good to move
'describe' into a subresource; it could return a general table format (csv?
xml? *shudder*) for convenient formatting as text, HTML, etc. We could
solve the 'get' in the same way-- we could have general parameters to the
'describe' page that cover things like "wide=true/false" and whether or not
to do a join with other resources (as describe does currently).

See kubernetes/dashboard#21 (comment)
<https://github.com/kubernetes/dashboard/issues/21#issuecomment-151634432>
and kubernetes/dashboard#32
<https://github.com/kubernetes/dashboard/pull/32>.

@smarterclayton <https://github.com/smarterclayton> Why do you call out pod
templates specifically?

generic commands that deal with object meta need to use runtime.Unstructured

Hopefully the pending client proposal will take care of this one.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17191/files#r44963839>.